File: python/numberofconnectedcomponentsinanundirectedpath.py
class Solution(object):
    
    def countComponents(self, n, edges):
        """
        :type n: int
        :type edges: List[List[int]]
        :rtype: int
        """
        import collections
        self.visited = set()
        count = 0
        self.edgesDict = collections.defaultdict(list)
        
        for e in edges:
            self.edgesDict[e[0]].append(e[1])
            self.edgesDict[e[1]].append(e[0])
        for i in xrange(n):
            if i not in self.visited:
                self.dfs(i)
                count += 1
                
        return count
        
    def dfs(self, j):
        self.visited.add(j)
        for e in self.edgesDict[j]:
            if e not in self.visited:
                self.dfs(e)
    
        

File: cpp/numberofconnectedcomponentsinanundirectedpath.cpp
class Solution {
public:
    int countComponents(int n, vector<pair<int, int>>& edges) {
        int count = 0;
        for (auto &e: edges) {
            adjList[e.first].push_front(e.second);
            adjList[e.second].push_front(e.first);
        }
        //cout << "countComponent" << endl;
        for (int i = 0; i < n; i++) {
            //cout << i << endl;
            if (vis.count(i) == 0) {
                count++;
                dfs(i);
            }
        }
        return count;
        
    }
    void dfs(int node) {
        if (vis.count(node) != 0) return;
        vis.insert(node);
        for (auto &n: adjList[node]) {
            //cout << node << " " << n << endl;
            dfs(n);
        }
    }
private:
    unordered_set<int> vis;
    unordered_map<int, list<int>> adjList;
};
